pkg/policy/api/utils.go

```diff
@@ -84,30 +52,28 @@ func (d *PortRuleDNS) Exists(rules L7Rules) bool {
 // Exists returns true if the L7 rule already exists in the list of rules
 func (h *PortRuleL7) Exists(rules L7Rules) bool {
 	for _, existingRule := range rules.L7 {
-		if h.Equal(existingRule) {
+		if h.DeepEqual(&existingRule) {
 			return true
 		}
 	}
 
 	return false
 }
 
-// Equal returns true if both rules are equal
-func (d *PortRuleDNS) Equal(o PortRuleDNS) bool {
-	return d != nil && d.MatchName == o.MatchName && d.MatchPattern == o.MatchPattern
+// DeepEqual returns true if both rules are equal
+func (d *PortRuleDNS) DeepEqual(o *PortRuleDNS) bool {
+	if d == nil {
+		return o == nil
+	}
+	return d.deepEqual(o)
 }
 
-// Equal returns true if both L7 rules are equal
-func (h *PortRuleL7) Equal(o PortRuleL7) bool {
-	if len(*h) != len(o) {
-		return false
-	}
-	for k, v := range *h {
-		if v2, ok := o[k]; !ok || v2 != v {
-			return false
-		}
+// DeepEqual returns true if both L7 rules are equal
+func (h *PortRuleL7) DeepEqual(o *PortRuleL7) bool {
+	if h == nil {
+		return o == nil || *o == nil
```

Comment:
Haven't checked, but do we need `return o == nil || *o == nil` in all `DeepEqual()` functions when `*o` is a map or a slice? Would `return o == nil || len(*o) == 0` have the same effect?

Reply:
I would consider two maps equal if they have the same elements, regardless the actual representation of an empty map, especially if that does not make a difference for the use of the map. But maybe for testing purposes we'd want DeepEqual to actually verify that the data structures have the same structure, and not only that they are functionally equivalent? On that note the code above is stating a `nil` pointer to a map `h` is DeepEqual to a non-nil pointer to a nil map `o`.

Reply:
@jrajahalme I took a look at go docs for the `reflect.DeepEqual`:
```
// Map values are deeply equal when all of the following are true:
// they are both nil or both non-nil, they have the same length,
// and either they are the same map object or their corresponding keys
// (matched using Go equality) map to deeply equal values.

// Slice values are deeply equal when all of the following are true:
// they are both nil or both non-nil, they have the same length,
// and either they point to the same initial entry of the same underlying array
// (that is, &x[0] == &y[0]) or their corresponding elements (up to length) are deeply equal.
// Note that a non-nil empty slice and a nil slice (for example, []byte{} and []byte(nil))
// are not deeply equal.
```

so I'll just perform the same behavior as the `reflect.DeepEqual`